examples/gossipsub-chat.rs

```diff
@@ -18,141 +18,151 @@
 //! cargo run --example gossipsub-chat --features=full
 //! ```
 //!
-//! It will print the [`PeerId`] and the listening addresses, e.g. `Listening on
-//! "/ip4/0.0.0.0/tcp/24915"`
-//!
-//! In the second terminal window, start a new instance of the example with:
+//! Mutual mDNS discovery may take a few seconds. When each peer does discover the other
+//! it will print a message like:
 //!
 //! ```sh
-//! cargo run --example gossipsub-chat --features=full -- /ip4/127.0.0.1/tcp/24915
+//! mDNS discovered a new peer: {peerId}
 //! ```
 //!
-//! The two nodes should then connect.
+//! Type a message and hit return: the message is sent and printed in the other terminal.
+//! Close with Ctrl-c.
+//!
+//! You can open more terminal windows and add more peers using the same line above.
+//!
+//! Once an additional peer is mDNS discovered it can participate in the conversation
+//! and all peers will receive messages sent from it.
+//!
+//! If a participant exits (Control-C or otherwise) the other peers will receive an mDNS expired
+//! event and remove the expired peer from the list of known peers.
 
 use async_std::io;
-use env_logger::{Builder, Env};
 use futures::{prelude::*, select};
 use libp2p::gossipsub::MessageId;
 use libp2p::gossipsub::{
-    GossipsubEvent, GossipsubMessage, IdentTopic as Topic, MessageAuthenticity, ValidationMode,
+    Gossipsub, GossipsubEvent, GossipsubMessage, IdentTopic as Topic, MessageAuthenticity,
+    ValidationMode,
+};
+use libp2p::{
+    gossipsub, identity,
+    mdns::{Mdns, MdnsConfig, MdnsEvent},
+    swarm::SwarmEvent,
+    NetworkBehaviour, PeerId, Swarm,
 };
-use libp2p::{gossipsub, identity, swarm::SwarmEvent, Multiaddr, PeerId};
 use std::collections::hash_map::DefaultHasher;
 use std::error::Error;
 use std::hash::{Hash, Hasher};
 use std::time::Duration;
 
 #[async_std::main]
 async fn main() -> Result<(), Box<dyn Error>> {
-    Builder::from_env(Env::default().default_filter_or("info")).init();
-
     // Create a random PeerId
     let local_key = identity::Keypair::generate_ed25519();
     let local_peer_id = PeerId::from(local_key.public());
-    println!("Local peer id: {:?}", local_peer_id);
+    println!("Local peer id: {}", local_peer_id);
 
-    // Set up an encrypted TCP Transport over the Mplex and Yamux protocols
+    // Set up an encrypted DNS-enabled TCP Transport over the Mplex protocol.
     let transport = libp2p::development_transport(local_key.clone()).await?;
 
+    // We create a custom network behaviour that combines Gossipsub and Mdns.
+    #[derive(NetworkBehaviour)]
+    struct MyBehaviour {
+        gossipsub: Gossipsub,
+        mdns: Mdns,
+    }
+
+    impl From<GossipsubEvent> for MyBehaviourEvent {
+        fn from(event: GossipsubEvent) -> Self {
+            MyBehaviourEvent::Gossipsub(event)
+        }
+    }
+
+    impl From<MdnsEvent> for MyBehaviourEvent {
+        fn from(event: MdnsEvent) -> Self {
+            MyBehaviourEvent::Mdns(event)
+        }
+    }
```

Comment:
You should be able to remove these `impl`s. They are provided through the `derive` macro above.